nixos/modules/services/x11/display-managers/default.nix

```
@@ -72,22 +77,19 @@ let
 
       # Speed up application start by 50-150ms according to
       # http://kdemonkey.blogspot.nl/2008/04/magic-trick.html
-      rm -rf "$HOME/.compose-cache"
-      mkdir "$HOME/.compose-cache"
+      compose_cache="''${XCOMPOSECACHE:-$HOME/.compose-cache}"
+      rm -rf "$compose_cache"
+      mkdir "$compose_cache"
+      unset compose_cache
```

I'd just make this
```suggestion
      mkdir -p "''${XCOMPOSECACHE:-$HOME/.compose-cache}"
```
and let the user worry about errors instead of unconditionally nuking whatever is already there.

Otherwise LGTM!

---

Thanks for reviewing!
The `rm` has been there since the time the trick was introduced in nixpkgs in 3b3dc83902a0d1c7735f112a8bd97bd5adaf8efb, and from 56e5b35f2e75925612268a2ee3fe6169abcf8517 it seems wiping the dir is actually necessary for it to work?

---

I see.  In that case I would recommend checking whether it is the correct directory first, à la
```suggestion
      compose_cache="''${XCOMPOSECACHE:-$HOME/.compose-cache}"
      [ -f "$compose_cache/file_that_should_be_there" ] && rm -rf "$compose_cache"
      mkdir -p "$compose_cache"
      unset compose_cache
```
In my opinion there should be some kind of safe-guard for bogus values of `XCOMPOSECACHE`.  You don't want anyone to accidentally set `XCOMPOSECACHE=$HOME` and have all their files vanish.

---

Yeah, that makes sense.
I looked into the internals of libX11 to see what file(s) are to be expected, in particular https://gitlab.freedesktop.org/xorg/lib/libx11/-/blob/master/modules/im/ximcp/imLcIm.c#L330 and at the cost of fragility it seems we could use
```
# sprintf (*res, "%s/%c%d_%03x_%08x_%08x", dir, _XimGetMyEndian(), XIM_CACHE_VERSION, (unsigned int)sizeof (DefTree), hash, hash2);
find ${XCOMPOSECACHE} -maxdepth 1 -regex '.*/[Bl][0-9]+_[0-9]+_[0-9a-f]+_[0-9a-f]+' -delete
```
and leave the directory itself alone. Should be pretty safe, but it's also rather magical/ugly and it could just stop working (although that part of the code is ancient and hasn't been changed in ages).

---

Or we could require that `$XCOMPOSECACHE` contains at least the string 'compose', which is less safe/universal but an easy heuristic?